docs: updated FAQ to accommodate new requirements prior to upgrading (#10434)
hashicorp/consul: website/content/docs/enterprise/license/faq.mdx

In previous versions of HashiCorp enterprise products, one server could distribute a license to other servers via the Raft protocol.
This will no longer work since each server must be able to find a valid license during the startup process.

## Q: What are the upgrade requirements?

> All customers on Consul Enterprise 1.8/1.9 must first upgrade their agents to the latest patch release.
> The license file must also be set in an environment variable or file path, otherwise the Consul agents will fail to retrieve the license with a valid agent token.

Reviewer: ```suggestion
All customers on Consul Enterprise 1.8/1.9 must first upgrade their client and server agents to the latest patch release.
During the upgrade the license file must also be configured on client agents in an environment variable or file path, otherwise the Consul agents will fail to retrieve the license with a valid agent token.
```

In previous versions of HashiCorp enterprise products, one server could distribute a license to other servers via the Raft protocol.
This will no longer work since each server must be able to find a valid license during the startup process.

## Q: What are the upgrade requirements?

All customers on Consul Enterprise 1.8/1.9 must first upgrade their agents to the latest patch release.
The license file must also be set in an environment variable or file path. Otherwise,
the Consul agents will fail to retrieve the license with a valid agent token.
The upgrade process varies if ACLs are enabled or disabled in the cluster. The following table and corresponding sections describe common 1.10.0 upgrade scenarios.

> 

Reviewer 2: @mkeeler has a whole new section on 1.10.0 upgrades in https://github.com/hashicorp/consul/pull/10446 that go over each scenario in detail. Rather than repeating them here, I think its best to share the most relevant bit of information which is the first paragraph above until "The upgrade process varies if ACLs are enabled or disabled in the cluster", and then link to the new page @mkeeler added to our upgrade section of the docs that's landing in that PR

Reviewer 2: Link to page from #10446 for 1.10.0 will be in `/docs/upgrading/instructions/upgrade-to-1-10-x`, to link to from here

Author: Great, we'll remove the table and the subsections and add the link.

| Client Version | Server Version | ACLs     | Guidance                                                                                                                                      |
| -------------- | -------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
| 1.9.x          | 1.9.x          | Enabled  | See [Upgrade Path for Consul Clients and Servers 1.9.x - ACLs Enabled](faq#upgrade-path-for-consul-clients-and-servers-1-9-x-acls-enabled).   |
| 1.9.x          | 1.9.x          | Disabled | See [Upgrade Path for Consul Clients and Servers 1.9.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-9-x-acls-disabled). |
| 1.8.x          | 1.8.x          | Enabled  | See [Upgrade Path for Consul Clients and Servers 1.8.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-8-x-acls-enabled).  |
| 1.8.x          | 1.8.x          | Disabled | See [Upgrade Path for Consul Clients and Servers 1.8.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-8-x-acls-disabled). |

> For Consul clients and servers older than 1.8, follow the instructions for [upgrading to specific versions](https://www.consul.io/docs/upgrading/upgrade-specific) to install Consul 1.8.X on clients and servers before performing one of the following upgrade scenarios.

Reviewer: ```suggestion
For Consul clients and servers older than 1.8, follow the instructions for [upgrading to specific versions](/docs/upgrading/upgrade-specific) to install Consul 1.8.X on clients and servers before performing one of the following upgrade scenarios.
```
minor nit: By convention most URLs on the same property use relative paths.

In previous versions of HashiCorp enterprise products, one server could distribute a license to other servers via the Raft protocol.
This will no longer work since each server must be able to find a valid license during the startup process.

## Q: What are the upgrade requirements?

> If ACLs are enabled, customers on Consul Enterprise 1.8+ or 1.9+ must upgrade Consul agents to the latest patch available (1.8.13 or 1.9.7) prior to upgrading the Consul server instances to 1.10.0.

Reviewer 2: per our talk earlier today and what I understood, *any* upgrade to 1.10.0 enterprise must start with upgrading to the latest patch, so I would replace this sentence with

"Any customers on Consul Enterprise 1.8/1.9 should first upgrade their agents to the latest patch release. Additionally, the license file must be set in an environment variable or file path, and the upgrade process varies if ACLs are not enabled in the cluster. We have provided additional instructions below"

cc @mkeeler for additional review on what I said above

Reviewer 3: That all looks correct to me.

| 1.9.x          | 1.9.x          | Enabled  | Do not upgrade Snapshot Agent to the latest 1.9.x. | See [Upgrade Path for Consul Clients and Servers 1.9.x - ACLs Enabled](faq#upgrade-path-for-consul-clients-and-servers-1-9-x-acls-enabled).   |
| 1.9.x          | 1.9.x          | Disabled | Snapshot Agent affected by upgrade.                | See [Upgrade Path for Consul Clients and Servers 1.9.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-9-x-acls-disabled). |
| 1.8.x          | 1.8.x          | Enabled  | Do not upgrade Snapshot Agent to the latest 1.8.x  | See [Upgrade Path for Consul Clients and Servers 1.8.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-8-x-acls-enabled).  |
| 1.8.x          | 1.8.x          | Disabled | Snapshot Agent affected by upgrade.                | See [Upgrade Path for Consul Clients and Servers 1.8.x - ACLs Disabled](faq#upgrade-path-for-consul-clients-and-servers-1-8-x-acls-disabled). |

> 

Reviewer 2: I would pull out the snapshot agent column out of this table. Its mentioned explicitly below for each section where its clear that when ACLs are disabled, we do require upgrading both the snapshot agent and clients/servers to the latest in the 1.9.x series.

## Q: Does this affect client agents?

For existing clusters, if the clients agents are using ACLs and have a valid token, then they will be able to retrieve the license from the server if the upgrade requirements have been met. Please see the [upgrade requirements](faq#q-what-are-the-upgrade-requirements).

> 

Reviewer: The upgrade requirements state that the license needs to be present on the host, yet this section states that the clients will be able to retrieve the license from the servers if they have a valid ACL token.

This is somewhat confusing, and I'm not sure how to re-word it.

Reviewer 4: I simplified the sentence and instead relying on the previous information provided at the top to avoid confusion.

```md
There are upgrade requirements that affect Consul Enterprise clients.
Please review the [upgrade requirements](faq#q-what-are-the-upgrade-requirements) documentation.
```

## Q: Does this affect client agents?

There are upgrade requirements that affect Consul Enterprise clients.
Please review the [upgrade requirements](faq#q-what-are-the-upgrade-requirements) documentation.

If the client agents are not using ACLs, then the client agents will be need to have the license on-disk (auto-loading) or as an environment variable.

> For new Consul clusters using Consul 1.10.0+ent, customers must ensure that there is a valid license on-disk (auto-loaded) or as an environment variable.

Reviewer 2: I think this second sentence "For new clusters using .." can be removed, since it doesn't apply to *new* 1.10.0 clients and this sentence is in the client agent section, also the latest commit added the upgrade requirements above and linked to that already..

> New Consul cluster deployments using 1.10.0+ent will need to have a valid license to successfully deploy.
> This valid license must be on-disk (auto-loaded) or as an environment variable.

Reviewer: ```suggestion
New Consul cluster deployments using 1.10.0+ent will need to have a valid license on servers to successfully deploy.
This valid license must be on-disk (auto-loaded) or as an environment variable.
```
